### PR TITLE
PS-707 Databox basket - Multi Part upload files to Expose

### PR DESCRIPTION
**Pull request by aynsix**

### CODE REVIEW

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 1]
```diff
@@ -149,13 +151,86 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+                'type' => $source->getType(),
+                'size' => (int)$source->getSize(),
+            ];
+
+            $resUploads = $this->create($config, $integrationToken)
+                ->request('POST', '/uploads', [
+                    'json' => $uploadsData,
+                ])
+                ->toArray()
+            ;
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:165]: Can be called after the multipart upload to "complete" upload

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 2]
```diff
@@ -149,13 +151,86 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+                    $resUploadPart = $this->create($config, $integrationToken)
+                        ->request('POST', '/uploads/'. $mUploadId .'/part', [
+                            'json' => ['part' => $partNumber],
+                        ])
+                        ->toArray()
+                    ;
+
+                    try {
+                        $headerPutPart = $this->uploadClient->request('PUT', $resUploadPart['url'], [
+                                'body' => fread($file, 10 * 1024 * 1024), // 10Mo
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:185]: Find a way to do not consume memory.
https://symfony.com/doc/current/http_client.html#uploading-data

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 3]
```diff
@@ -149,13 +151,86 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+                            ])->getHeaders()
+                        ;
+
+                    } catch (\Exception $e) {
+                        sleep (1); // retry after 1 second
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:190]: No sleep, better to retry in another event

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 4]
```diff
@@ -149,13 +151,86 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+                        try {
+                            $headerPutPart = $this->uploadClient->request('PUT', $resUploadPart['url'], [
+                                'body' => fread($file, 10 * 1024 * 1024), // 10Mo
+                            ])->getHeaders()
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:194]: Should be factorized and retried recusrsively

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 5]
```diff
@@ -149,13 +151,86 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+                        ;
+                        } catch (\Exception $e) {
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:196]:
```suggestion
                        } catch (\Throwable $e) {
```
Please check others!

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 6]
```diff
@@ -149,13 +151,86 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+                    ];
+
+                    $partNumber++;
+                }
+                
+                $isPartsComplete = true;
+                fclose($file);
+            } catch (\Exception $e) {
+                $this->create($config, $integrationToken)
+                    ->request('DELETE', '/uploads/'. $mUploadId);
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:213]:
```suggestion
                $this->create($config, $integrationToken)
                    ->request('DELETE', '/uploads/'. $mUploadId);
                 throw $e;
```

File: lib/php/storage-bundle/Controller/MultipartUploadCancelAction.php [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    {
+    }
+    
+    public function __invoke(MultipartUpload $data, Request $request)
+    {
+        try {
+            $this->uploadManager->cancelMultipartUpload($data->getPath(), $data->getUploadId());
+        } catch (\Exception $e) {
+            
+        }
```
Comment (4rthem) [inline lib/php/storage-bundle/Controller/MultipartUploadCancelAction.php:25]:
```suggestion
        } catch (\Throwable $e) {
            // S3 storage will clean up its uncomplete uploads automatically
        }
```

File: lib/php/storage-bundle/Controller/MultipartUploadCompleteAction.php [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Alchemy\StorageBundle\Upload\UploadManager;
+use Alchemy\StorageBundle\Entity\MultipartUpload;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+
+final class MultipartUploadCompleteAction extends AbstractController
```
Comment (4rthem) [inline lib/php/storage-bundle/Controller/MultipartUploadCompleteAction.php:15]: Can be removed

File: lib/php/storage-bundle/Entity/MultipartUpload.php [hunk 9]
```diff
@@ -67,7 +69,42 @@
                         ]],
                 ],
             ]),
-        new Delete(openapiContext: ['summary' => 'Cancel an upload', 'description' => 'Cancel an upload.']),
+        new Post(
```
Comment (4rthem) [inline lib/php/storage-bundle/Entity/MultipartUpload.php:72]: Can be removed

File: lib/php/storage-bundle/Resources/config/services.yaml [hunk 10]
```diff
@@ -42,6 +42,14 @@ services:
         tags:
             - { name: controller.service_arguments }
 
+    Alchemy\StorageBundle\Controller\MultipartUploadCancelAction:
+        tags:
+            - { name: controller.service_arguments }        
+
+    Alchemy\StorageBundle\Controller\MultipartUploadCompleteAction:
+        tags:
+            - { name: controller.service_arguments }
```
Comment (4rthem) [inline lib/php/storage-bundle/Resources/config/services.yaml:51]: rm

**Review by 4rthem — CHANGES_REQUESTED** [review]

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 11]
```diff
@@ -207,4 +254,21 @@ public function deleteAsset(IntegrationConfig $config, IntegrationToken $integra
...
+    {
+        if ($retryCount > 0) {
+            $retryCount--;
+            try {
+                return $this->uploadClient->request('PUT', $url, [
+                    'body' => fread($handleFile, $partSize),
+                ])->getHeaders();
+            } catch (\Throwable $e) {
+                if ($retryCount == 0) {
+                    throw $e; // retry unsuccess
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:268]: Useless comment

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 12]
```diff
@@ -207,4 +254,21 @@ public function deleteAsset(IntegrationConfig $config, IntegrationToken $integra
             ->request('DELETE', '/assets/'.$assetId)
         ;
     }
+
+    private function putPart(string $url, mixed $handleFile, int $partSize, int $retryCount)
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:258]:
```suggestion
    private function putPart(string $url, mixed $handleFile, int $partSize, int $retryCount): void
```

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 13]
```diff
@@ -207,4 +254,21 @@ public function deleteAsset(IntegrationConfig $config, IntegrationToken $integra
...
+    {
+        if ($retryCount > 0) {
+            $retryCount--;
+            try {
+                return $this->uploadClient->request('PUT', $url, [
+                    'body' => fread($handleFile, $partSize),
+                ])->getHeaders();
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:265]: Still TODO: try to use closure

**Review by 4rthem — CHANGES_REQUESTED** [review]

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php (hunk 13 above)
Comment (aynsix) [reply databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:265]: fread always return false on the body closure

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 14]
```diff
@@ -127,35 +129,121 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
         $source = $asset->getSource();
 
         $fetchedFilePath = $this->fileFetcher->getFile($source);
+        $fileSize =  filesize($fetchedFilePath);
+
+        // allowed part size is up to 5Mo https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html
+        $partSize = 10 * 1024 * 1024; // 10Mo
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:135]:
```suggestion
        // @see https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html
        $partSize = 100 * 1024 * 1024; // 100Mb
```

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 15]
```diff
@@ -127,35 +129,121 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+                        'json' => $uploadsData,
+                    ])
+                    ->toArray()
+                ;
+    
+                $mUploadId = $resUploads['id'];
+    
+                $parts['Parts'] = [];
+    
+                // Upload the file in parts.
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:156]: useless comment

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 16]
```diff
@@ -127,35 +129,121 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+                $exposeAssetId = $pubAsset['id'];
+    
+                $this->uploadClient->request('PUT', $pubAsset['uploadURL'], [
+                    'headers' => [
+                        'Content-Type' => $source->getType(),
+                        'Content-Length' => filesize($fetchedFilePath),
+                    ],
+                    'body' => fopen($fetchedFilePath, 'r'),
+                ]);
+            }
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:246]: Is this the old upload way?

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 17]
```diff
@@ -127,35 +129,121 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+                    ],
+                ], $extraData);
+    
+                $pubAsset = $this->create($config, $integrationToken)
+                    ->request('POST', '/assets', [
+                        'json' => $data,
+                    ])
+                    ->toArray()
+                ;
+                $exposeAssetId = $pubAsset['id'];
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:237]: unused

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 18]
```diff
@@ -127,35 +129,121 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+                ], $extraData);
+    
+                $pubAsset = $this->create($config, $integrationToken)
+                    ->request('POST', '/assets', [
+                        'json' => $data,
+                    ])
+                    ->toArray()
+                ;
+    
+                $exposeAssetId = $pubAsset['id'];
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:216]: unused

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 19]
```diff
@@ -127,35 +129,121 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+
         try {
-            $data = array_merge([
-                'publication_id' => $publicationId,
-                'asset_id' => $asset->getId(),
-                'title' => $resolvedTitle,
-                'description' => $description,
-                'translations' => $translations,
-                'upload' => [
+            if ($fileSize >= $partSize ) {
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:138]: No need to condition this. Multipart upload should work with one part as well.

**Review by 4rthem — CHANGES_REQUESTED** [review]

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 20]
```diff
@@ -127,17 +129,78 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+                ])
+                ->toArray()
+            ;
+
+            $mUploadId = $resUploads['id'];
+
+            $parts['Parts'] = [];
+
+            try {
+                $file = fopen($fetchedFilePath, 'r');
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:156]:
```suggestion
                $fd = fopen($fetchedFilePath, 'r');
```

File: databox/api/src/Integration/Phrasea/Expose/ExposeClient.php [hunk 21]
```diff
@@ -127,17 +129,78 @@ public function postAsset(IntegrationConfig $config, IntegrationToken $integrati
...
+
+                while ( ($fileSize - $alreadyUploaded) > 0) {
+                    $resUploadPart = $this->create($config, $integrationToken)
+                        ->request('POST', '/uploads/'. $mUploadId .'/part', [
+                            'json' => ['part' => $partNumber],
+                        ])
+                        ->toArray()
+                    ;
+
+                    if (($fileSize - $alreadyUploaded) <= 2*$partSize) {
```
Comment (4rthem) [inline databox/api/src/Integration/Phrasea/Expose/ExposeClient.php:171]: Why `2*` ?

```suggestion
                    if (($fileSize - $alreadyUploaded) < $partSize) {
```

**Review by 4rthem — CHANGES_REQUESTED** [review]

**Pull request merged**